src/nix-build/nix-build.cc

```diff
@@ -317,7 +338,8 @@ static void main_nix_build(int argc, char * * argv)
 
                     auto sourcePath = lookupFileArg(*state,
                                     baseDir);
-                    auto resolvedPath = resolveExprPath(sourcePath);
+                    auto resolvedPath =
+                        isNixShell ? resolveShellExprPath(sourcePath) : resolveExprPath(sourcePath);
```

Comment:
Maybe the condition should be: `isNixShell` _and_ no `-A` was passed?

- https://github.com/NixOS/nix/issues/11699